python/pyspark/ml/util.py

```diff
@@ -52,3 +71,133 @@ def _randomUID(cls):
+
+    @since("2.0.0")
+    def save(self, path):
+        """Saves the ML instances to the input path."""
+        self._jwrite.save(path)
+
+    @since("2.0.0")
+    def overwrite(self):
+        """Overwrites if the output path already exists."""
+        self._jwrite.overwrite()
+        return self
+
+    @since("2.0.0")
+    def context(self, sqlContext):
+        """Sets the SQL context to use for saving."""
+        self._jwrite.context(sqlContext._ssql_ctx)
+        return self
+
+
+@inherit_doc
+class MLWritable(object):
+    """
+    .. note:: Experimental
+
+    Mixin for ML instances that provide MLWriter through their Scala
+    implementation.
+
+    .. versionadded:: 2.0.0
+    """
+
+    @since("2.0.0")
+    def write(self):
+        """Returns an MLWriter instance for this ML instance."""
+        return MLWriter(self)
+
+    @since("2.0.0")
+    def save(self, path):
+        """Save this ML instance to the given path, a shortcut of `write().save(path)`."""
+        if not isinstance(path, basestring):
+            raise TypeError("path should be a basestring, got type %s" % type(path))
+        self.write().save(path)
+
+
+@inherit_doc
+class MLReader(object):
+    """
+    .. note:: Experimental
+
+    Utility class that can load ML instances.
+
+    .. versionadded:: 2.0.0
+    """
+
+    def __init__(self, instance):
+        self._instance = instance
+        self._jread = instance._java_obj.read()
+
+    @since("2.0.0")
+    def load(self, path):
+        """Loads the ML component from the input path."""
+        java_obj = self._jread.load(path)
+        self._instance._java_obj = java_obj
+        self._instance.uid = java_obj.uid()
+        self._instance._transfer_params_from_java(True)
+        return self._instance
+
+    @since("2.0.0")
+    def context(self, sqlContext):
+        """Sets the SQL context to use for loading."""
+        self._jread.context(sqlContext._ssql_ctx)
+        return self
+
+
+@inherit_doc
+class MLReadable(object):
+    """
+    .. note:: Experimental
+
+    Mixin for objects that provide MLReader using its Scala implementation.
+
+    .. versionadded:: 2.0.0
+    """
+
+    @classmethod
+    def _java_loader_class(cls):
+        """
+        Returns the full class name of the Java loader. The default
+        implementation replaces "pyspark" by "org.apache.spark" in
+        the Python full class name.
+        """
+        java_package = cls.__module__.replace("pyspark", "org.apache.spark")
+        return ".".join([java_package, cls.__name__])
+
+    @classmethod
+    def _load_java_obj(cls):
+        """Load the peer Java object."""
+        java_class = cls._java_loader_class()
+        java_obj = _jvm()
+        for name in java_class.split("."):
+            java_obj = getattr(java_obj, name)
+        return java_obj
+
+    @classmethod
+    @since("2.0.0")
+    def read(cls):
+        """Returns an MLReader instance for this class."""
+        instance = cls()
```

Comment:
This could be done within the MLReader init method (if we want to make MLReadable more general and not JVM-specific).